### ISSUE TITLE
autosuggest behavior has weird artifacts inside emacs ansi-term mode

### ISSUE THREAD

**Issue by krader1961** [fish-shell/fish-shell#3282]
Issue #1363 was opened to discuss the desirability of providing a mechanism for disabling autosuggestions. [Recently](https://github.com/fish-shell/fish-shell/issues/1363#issuecomment-236954788) user @galbacarys pointed out that the autosuggestion feature produces weird artifacts when fish is run inside emacs ansi-term mode.

I was able to reproduce this on macOS 10.11.6 running emacs 24.5.1 and fish 2.3.1 inside iTerm2 3.0.

![screenshot 2016-08-02 19 44 41](https://cloud.githubusercontent.com/assets/7545917/17352216/a4f07cbe-58e9-11e6-91d9-bf174b055563.png)

TBD: Is this a bug in emacs ansi-term? Its terminfo (`eterm-color`) definition? Or is fish doing something that happens to work in other terminals but isn't 100% valid? Note that the "4m" you see in that screenshot is from the sequence for enabling character underlining (`\e[4m`) which the `eterm-color` terminfo definition claims to support. And it does seem to support it in as much as some chars are underlined. The question is what is causing that otherwise valid escape sequence to not be correctly interpreted by emacs ansi-term?

**Comment by floam** [comment]
Are you sure OS X has a terminfo for eterm-color?

**Comment by floam** [comment]
(Not the one emacs comes with, but in /usr/share/terminfo)

**Comment by krader1961** [comment]
> Are you sure OS X has a terminfo for eterm-color?

It does not. It does have one for `eterm`. Running `infocmp -I eterm-color` within emacs says it is reconstructing it from termcap. Doing the same thing outside of emacs fails with "infocmp: couldn't open terminfo file (null)." TBD is the relevance of that observation. In other words, is fish using the magically constructed terminfo analog to whatever termcap definition emacs has created?

**Comment by floam** [comment]
On my system:

```
set -x TERM eterm-info; fish
<E> fish: Could not set up terminal
<E> fish: TERM environment variable set to 'eterm-info'
<E> fish: Check that this terminal type is supported on this system
<E> fish: Using fallback terminal type 'ansi'
```

How did you interrogate the terminfo definition to determine that it supports underline for `eterm-color`?

**Comment by krader1961** [comment]
> How did you interrogate the terminfo definition to determine that it supports underline for eterm-color?
![screenshot 2016-08-02 20 43 01](https://cloud.githubusercontent.com/assets/7545917/17353137/c4130fdc-58f1-11e6-959f-014012807da9.png)

**Comment by floam** [comment]
What does this do?

`tic /usr/share/emacs/22.1/etc/e/eterm-color.ti `

**Comment by floam** [comment]
Does fish perhaps not do a setupterm() again after whatever is getting set in the environment?

**Comment by floam** [comment]
I know for certain that fish has some real problems and does not do things it should if I change $TERM willy nilly, I'd guess it's related.

**Comment by krader1961** [comment]
> Does fish perhaps not do a setupterm() again after whatever is getting set in the environment?

I don't understand your question. In this specific instance `$TERM` is not being changed on the fly. It is already set when fish starts running. So even if there is a bug in handling `$TERM` changing on the fly (something I thought I fixed with commit eb834f47efeca3c2100163e892ab41df670b0687 on 2016-06-01) that wouldn't seem to be relevant to this issue as I was able to recreate the problem with explicitly setting `TERM`.

**Comment by floam** [comment]
If $TERMCAP/$TERMINFO changes on the fly fish will react?

**Comment by krader1961** [comment]
> If $TERMCAP/$TERMINFO changes on the fly fish will react?

Yes to `$TERMINFO`, no to `$TERMCAP`. Although support for dynamically changing the former may be incomplete. See `curses_variable` in *src/env.cpp*. Regardless, I fail to see the relevance of this line of investigation given that the problem can be reproduced with a fresh fish shell where those vars are already set when the shell is started.

![screenshot 2016-08-02 21 55 52](https://cloud.githubusercontent.com/assets/7545917/17354276/f6c0fd54-58fb-11e6-98e8-6173c7195311.png)

**Comment by floam** [comment]
OK. I take it tic putting that into ~/.terminfo had no effect. Beats me.

**Comment by krader1961** [comment]
I'm 100% certain the problem is a bug in emacs ansi-term mode. Specifically its termcap definition is broken. Notice that `infocmp -I eterm-color` reports `smul=\E[4m` and `rmul=\E24m`. Notice the lack of a `[` (left-bracket) character after the escape character in the `rmul` sequence. You can see the same thing in the `echo $TERMCAP` output by looking at the `us` and `ue` sequences (which are identical to the terminfo `smul` and `rmul` sequences). If the missing left-bracket were present in the `ue` or `rmul` sequence it would probably work as expected.

**Comment by krader1961** [comment]
FWIW, doing a google search for `emacs ansi-term underline` returns results that suggests this issue affects other programs such as [zsh](http://emacs.stackexchange.com/questions/7372/stray-trailing-4m-before-prompt-with-zsh-in-m-x-ansi-term).

I should have mentioned in my previous comment that `\E[24m` is a valid ANSI SGR (set graphic rendition sequence) to [reset underline mode](http://ttssh2.osdn.jp/manual/en/about/ctrlseq.html). Without the left bracket it is meaningless; at least with respect to terminals adhering to the ANSI X3.64 standard.

I am strongly inclined to close this as a bug in Emacs that fish should not attempt to work around.

**Comment by floam** [comment]
Sorry, bumped the button.

**Comment by faho** [comment]
As a note, I cannot reproduce this on linux. My system ncurses ships with the correct terminfo.

>I am strongly inclined to close this as a bug in Emacs that fish should not attempt to work around.

I agree. Note that eterm has more issues than this - it's got a whopping 8 colors, it will barf when we try to set the title. Currently, I cannot call eterm a good terminal. It's in need of some serious love.

This has come up before, btw: #2411 is the same thing. There it was reported that `tic -o ~/.terminfo $TERMINFO/e/eterm-color.ti` would fix it. And now we know why this only occurs on macOS - it's shipping a borken definition.

**Comment by krader1961** [comment]
Actually it's a dup of #2441, not #2411, and also #2983. I'd be willing to add a workaround for this specific problem but as @faho noted emacs ansi-term has other issues. Also, the user can work around this by using their own copy of a correct terminfo file for eterm-color.

**Comment by galbacarys** [comment]
The fix in #2441 worked correctly, thanks!
